File: src/eye/scaling.py
"""Coordinate scaling utilities for sending smaller screenshots to the LLM.

When a screenshot is captured at high logical resolution (e.g. 1728x1117),
sending it directly to the LLM wastes tokens and can reduce click precision
because the model must reason over a larger coordinate space.

This module provides:
- :func:`find_target_resolution` — pick a standard resolution that matches
  the screen's aspect ratio.
- :func:`scale_screenshot` — resize a :class:`Screenshot` for the LLM,
  preserving the original screen dimensions for reverse mapping.
- :func:`scale_coordinates` — map model-space coordinates back to screen space.
"""

from __future__ import annotations

import base64
import io
import logging
from typing import Any

from PIL import Image

from src.eye.base import Screenshot

logger = logging.getLogger(__name__)

# Standard target resolutions (width, height).
# Chosen to cover common monitor aspect ratios.
SCALING_TARGETS: list[tuple[int, int]] = [
    (1024, 768),   # XGA   — 4:3   (1.333)
    (1280, 800),   # WXGA  — 16:10 (1.600)
    (1366, 768),   # FWXGA — ~16:9 (1.779)
]

ASPECT_TOLERANCE = 0.05  # 5%


def find_target_resolution(
    width: int,
    height: int,
    match_mode: str = "aspect_ratio",
) -> tuple[int, int] | None:
    """Return the best standard target for the given screen dimensions.

    Parameters:
        width: Source image width.
        height: Source image height.
        match_mode: ``"aspect_ratio"`` (default) picks the target whose
            aspect ratio is closest within ``ASPECT_TOLERANCE``.
            ``"pixel_count"`` picks the target whose total pixel count
            is closest to the source (still only from targets smaller
            than the source).

    Returns:
        ``None`` if no target qualifies or the image is already small.
    """
    if width <= 0 or height <= 0:
        return None

    # Filter to targets strictly smaller than the source.
    candidates = [
        (tw, th) for tw, th in SCALING_TARGETS
        if not (tw >= width and th >= height)
    ]
    if not candidates:
        return None

    if match_mode == "pixel_count":
        src_pixels = width * height
        best: tuple[int, int] | None = None
        best_diff = float("inf")
        for tw, th in candidates:
            diff = abs(src_pixels - tw * th)
            if diff < best_diff:
                best = (tw, th)
                best_diff = diff
        return best

    # Default: aspect_ratio matching.
    aspect = width / height
    best_ar: tuple[int, int] | None = None
    best_ar_diff = float("inf")

    for tw, th in candidates:
        target_aspect = tw / th
        diff = abs(aspect - target_aspect) / aspect
        if diff < ASPECT_TOLERANCE and diff < best_ar_diff:
            best_ar = (tw, th)
            best_ar_diff = diff

    return best_ar


def scale_screenshot(screenshot: Screenshot, target: tuple[int, int]) -> Screenshot:
    """Resize *screenshot* to *target* resolution for the LLM.

    The returned :class:`Screenshot` carries the original screen dimensions in
    ``screen_width`` / ``screen_height`` so that coordinates from the LLM can
    be mapped back.

    When the source :class:`Screenshot` carries an ``image`` (PIL Image), we
    resize from that lossless source to avoid double WebP compression.
    """
    tw, th = target
    if screenshot.width == tw and screenshot.height == th:
        return screenshot

    # Prefer the retained PIL Image (lossless) over base64-decode (lossy).
    if screenshot.image is not None:
        img = screenshot.image.resize((tw, th), Image.Resampling.LANCZOS)
    else:
        img_data = base64.b64decode(screenshot.base64)
        img = Image.open(io.BytesIO(img_data))
        img = img.resize((tw, th), Image.Resampling.LANCZOS)

    buf = io.BytesIO()
    img.save(buf, format="WEBP", lossless=True)
    b64 = base64.b64encode(buf.getvalue()).decode("ascii")

    logger.info(
        "Scaled screenshot %dx%d -> %dx%d (b64_len %d -> %d)",
        screenshot.width, screenshot.height, tw, th,
        len(screenshot.base64), len(b64),
    )

    return Screenshot(
        base64=b64,
        width=tw,
        height=th,
        scale_factor=screenshot.scale_factor,
        mime_type="image/webp",
        screen_width=screenshot.width,
        screen_height=screenshot.height,
        image=img,
    )


def scale_coordinates(
    x: int | Any,
    y: int | Any,
    model_width: int,
    model_height: int,
    screen_width: int,
    screen_height: int,
) -> tuple[int, int]:
    """Map coordinates from model (scaled) space back to screen space.

    *x* and *y* are coerced to ``int`` so that string values returned by
    some LLMs (e.g. ``"640"``) are handled gracefully.  Raises
    :class:`ValueError` if conversion fails.
    """
    x = int(x)
    y = int(y)
    sx = round(x * screen_width / model_width)
    sy = round(y * screen_height / model_height)
    return sx, sy


def scale_tool_args(
    tool_name: str,
    args: dict[str, Any],
    model_width: int,
    model_height: int,
    screen_width: int,
    screen_height: int,
) -> dict[str, Any]:
    """Return a copy of *args* with coordinates scaled to screen space.

    Only tools that use screen coordinates (``click``, ``drag``, ``scroll``)
    are affected.  Other tools pass through unchanged.
    """
    if tool_name == "click":
        sx, sy = scale_coordinates(
            args["x"], args["y"],
            model_width, model_height, screen_width, screen_height,
        )
        return {**args, "x": sx, "y": sy}

    if tool_name == "drag":
        sx1, sy1 = scale_coordinates(
            args["start_x"], args["start_y"],
            model_width, model_height, screen_width, screen_height,
        )
        sx2, sy2 = scale_coordinates(
            args["end_x"], args["end_y"],
            model_width, model_height, screen_width, screen_height,
        )
        return {
            **args,
            "start_x": sx1, "start_y": sy1,
            "end_x": sx2, "end_y": sy2,
        }

    if tool_name == "scroll":
        sx, sy = scale_coordinates(
            args["x"], args["y"],
            model_width, model_height, screen_width, screen_height,
        )
        return {**args, "x": sx, "y": sy}

    return args

File: see-agent-corp/src/eye/scaling.rs
use base64::Engine;

use crate::consts::{ASPECT_TOLERANCE, SCALE_TARGETS};
use crate::error::{Result, CorpError};

use super::capture::Screenshot;

/// Find the best target resolution for scaling a screenshot.
///
/// Returns `None` if the image is already small enough (no scaling needed).
///
/// `match_mode`: "aspect_ratio" (default) or "pixel_count"
pub fn find_target_resolution(
    width: u32,
    height: u32,
    match_mode: &str,
) -> Option<(u32, u32)> {
    if width == 0 || height == 0 {
        return None;
    }

    // Filter: only targets where the source exceeds at least one dimension
    let candidates: Vec<(u32, u32)> = SCALE_TARGETS
        .iter()
        .copied()
        .filter(|&(tw, th)| !(tw >= width && th >= height))
        .collect();

    if candidates.is_empty() {
        return None;
    }

    match match_mode {
        "pixel_count" => {
            let src_pixels = width as u64 * height as u64;
            candidates
                .into_iter()
                .min_by_key(|&(tw, th)| {
                    let target_pixels = tw as u64 * th as u64;
                    src_pixels.abs_diff(target_pixels)
                })
        }
        _ => {
            // aspect_ratio mode (default)
            let aspect = width as f64 / height as f64;
            let mut best: Option<(u32, u32, f64)> = None;

            for (tw, th) in candidates {
                let target_aspect = tw as f64 / th as f64;
                let diff = (aspect - target_aspect).abs() / aspect;
                if diff < ASPECT_TOLERANCE
                    && (best.is_none() || diff < best.unwrap().2)
                {
                    best = Some((tw, th, diff));
                }
            }

            best.map(|(tw, th, _)| (tw, th))
        }
    }
}

/// Scale a screenshot to the target resolution.
///
/// Returns a new Screenshot with `screen_width`/`screen_height` set to
/// the original dimensions (for coordinate reverse-mapping).
pub fn scale_screenshot(
    screenshot: &Screenshot,
    target_w: u32,
    target_h: u32,
) -> Result<Screenshot> {
    if screenshot.width == target_w && screenshot.height == target_h {
        return Ok(screenshot.clone());
    }

    // Load image from retained bytes or decode from base64
    let img = if let Some(ref bytes) = screenshot.image_bytes {
        image::load_from_memory(bytes).map_err(|e| CorpError::Agent {
            message: format!("image decode error: {e}"),
        })?
    } else {
        let bytes = base64::engine::general_purpose::STANDARD
            .decode(&screenshot.base64)
            .map_err(|e| CorpError::Agent {
                message: format!("base64 decode error: {e}"),
            })?;
        image::load_from_memory(&bytes).map_err(|e| CorpError::Agent {
            message: format!("image decode error: {e}"),
        })?
    };

    // Resize with Lanczos
    let resized = img.resize_exact(target_w, target_h, image::imageops::FilterType::Lanczos3);

    // Encode to WebP
    let mut webp_bytes = Vec::new();
    let mut cursor = std::io::Cursor::new(&mut webp_bytes);
    resized
        .write_to(&mut cursor, image::ImageFormat::WebP)
        .map_err(|e| CorpError::Agent {
            message: format!("webp encode error: {e}"),
        })?;

    let b64 = base64::engine::general_purpose::STANDARD.encode(&webp_bytes);

    Ok(Screenshot {
        base64: b64,
        width: target_w,
        height: target_h,
        scale_factor: screenshot.scale_factor,
        mime_type: "image/webp".to_owned(),
        screen_width: Some(screenshot.screen_width.unwrap_or(screenshot.width)),
        screen_height: Some(screenshot.screen_height.unwrap_or(screenshot.height)),
        image_bytes: Some(webp_bytes),
    })
}

/// Reverse-map coordinates from LLM (model) space back to screen space.
///
/// `(x, y)` = coordinates the LLM returned
/// `model_w, model_h` = the dimensions the LLM saw (scaled screenshot)
/// `screen_w, screen_h` = the original logical screen dimensions
pub fn scale_coordinates(
    x: i32,
    y: i32,
    model_w: u32,
    model_h: u32,
    screen_w: u32,
    screen_h: u32,
) -> (i32, i32) {
    if model_w == 0 || model_h == 0 {
        return (x, y);
    }
    let sx = (x as f64 * screen_w as f64 / model_w as f64).round() as i32;
    let sy = (y as f64 * screen_h as f64 / model_h as f64).round() as i32;
    (sx, sy)
}

/// Scale coordinate fields in tool arguments back to screen space.
///
/// Applies `scale_coordinates` to the relevant fields for click, drag, scroll.
pub fn scale_tool_args(
    tool_name: &str,
    args: &mut serde_json::Value,
    model_w: u32,
    model_h: u32,
    screen_w: u32,
    screen_h: u32,
) {
    match tool_name {
        "click" | "scroll" => {
            if let (Some(x), Some(y)) = (
                args.get("x").and_then(|v| v.as_i64()).map(|v| v as i32),
                args.get("y").and_then(|v| v.as_i64()).map(|v| v as i32),
            ) {
                let (sx, sy) = scale_coordinates(x, y, model_w, model_h, screen_w, screen_h);
                args["x"] = serde_json::json!(sx);
                args["y"] = serde_json::json!(sy);
            }
        }
        "drag" => {
            if let (Some(sx), Some(sy)) = (
                args.get("start_x")
                    .and_then(|v| v.as_i64())
                    .map(|v| v as i32),
                args.get("start_y")
                    .and_then(|v| v.as_i64())
                    .map(|v| v as i32),
            ) {
                let (nsx, nsy) = scale_coordinates(sx, sy, model_w, model_h, screen_w, screen_h);
                args["start_x"] = serde_json::json!(nsx);
                args["start_y"] = serde_json::json!(nsy);
            }
            if let (Some(ex), Some(ey)) = (
                args.get("end_x")
                    .and_then(|v| v.as_i64())
                    .map(|v| v as i32),
                args.get("end_y")
                    .and_then(|v| v.as_i64())
                    .map(|v| v as i32),
            ) {
                let (nex, ney) = scale_coordinates(ex, ey, model_w, model_h, screen_w, screen_h);
                args["end_x"] = serde_json::json!(nex);
                args["end_y"] = serde_json::json!(ney);
            }
        }
        _ => {} // Other tools: pass through unchanged
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn find_target_aspect_ratio() {
        // 1728x1117 (Retina MacBook logical) → aspect 1.547
        // Closest within 5%: (1280, 800) ratio=1.600, diff=3.4%
        let result = find_target_resolution(1728, 1117, "aspect_ratio");
        assert_eq!(result, Some((1280, 800)));
    }

    #[test]
    fn find_target_4_3_display() {
        // 1920x1440 → aspect 1.333 → matches (1024, 768) exactly
        let result = find_target_resolution(1920, 1440, "aspect_ratio");
        assert_eq!(result, Some((1024, 768)));
    }

    #[test]
    fn find_target_pixel_count() {
        let result = find_target_resolution(1728, 1117, "pixel_count");
        assert!(result.is_some());
    }

    #[test]
    fn find_target_already_small() {
        // Image smaller than all targets → no scaling
        let result = find_target_resolution(800, 600, "aspect_ratio");
        assert!(result.is_none());
    }

    #[test]
    fn find_target_zero_dims() {
        assert!(find_target_resolution(0, 0, "aspect_ratio").is_none());
        assert!(find_target_resolution(1920, 0, "aspect_ratio").is_none());
    }

    #[test]
    fn scale_coordinates_basic() {
        // LLM sees 1280x800, screen is 1728x1117
        let (sx, sy) = scale_coordinates(640, 400, 1280, 800, 1728, 1117);
        assert_eq!(sx, 864); // 640 * 1728 / 1280 = 864
        assert_eq!(sy, 559); // 400 * 1117 / 800 = 558.5 → 559
    }

    #[test]
    fn scale_coordinates_identity() {
        let (sx, sy) = scale_coordinates(100, 200, 1280, 800, 1280, 800);
        assert_eq!(sx, 100);
        assert_eq!(sy, 200);
    }

    #[test]
    fn scale_coordinates_zero_model() {
        let (sx, sy) = scale_coordinates(100, 200, 0, 0, 1280, 800);
        assert_eq!(sx, 100);
        assert_eq!(sy, 200);
    }

    #[test]
    fn scale_tool_args_click() {
        let mut args = json!({"x": 640, "y": 400});
        scale_tool_args("click", &mut args, 1280, 800, 1728, 1117);
        assert_eq!(args["x"], 864);
        assert_eq!(args["y"], 559);
    }

    #[test]
    fn scale_tool_args_drag() {
        let mut args = json!({"start_x": 100, "start_y": 100, "end_x": 640, "end_y": 400});
        scale_tool_args("drag", &mut args, 1280, 800, 1728, 1117);
        assert_eq!(args["end_x"], 864);
    }

    #[test]
    fn scale_tool_args_passthrough() {
        let mut args = json!({"command": "ls"});
        let original = args.clone();
        scale_tool_args("shell", &mut args, 1280, 800, 1728, 1117);
        assert_eq!(args, original);
    }

    #[test]
    fn find_target_1920x1080() {
        // 1920x1080 → aspect 1.778
        // Closest: (1366, 768) ratio=1.779, diff < 0.1%
        let result = find_target_resolution(1920, 1080, "aspect_ratio");
        assert_eq!(result, Some((1366, 768)));
    }

    #[test]
    fn find_target_2560x1600() {
        // 2560x1600 → aspect 1.600
        // Exact match: (1280, 800) ratio=1.600
        let result = find_target_resolution(2560, 1600, "aspect_ratio");
        assert_eq!(result, Some((1280, 800)));
    }
}
